### PR TITLE
mgr/telemetry: add some more telemetry

### PR DESCRIPTION
**Pull request by liewegas**
I'm looking at the user survey and trying to add anything we're asking about:

- [x] num file systems
- [x] cephfs data pools
- [x] cephfs active mds count (+ more)
- [x] cephfs file count
- [x] cephfs byte count
- [x] cephfs snapshots used or not
- [x] mds cache size
- [x] ipv4 vs ipv6 (for mons...)
- [x] msgr v1 vs msgr v2 (for mons...)
- [x] rgw realm count, zonegroup count, zone count
- [ ] rgw auth mechanism(s) used
- [ ] rgw apis used (s3 vs swift)
- [ ] rbd snapshots used
- [x] rbd num volumes?
- [x] rbd mirroring used
- [x] osds using dedicated cluster_network
- [x] pool: which EC codes are used
- [x] pool: which EC plugins are used
- [x] pool: cache tiering enabled
- [x] host count
- [ ] osds per node??

### CODE REVIEW

File: src/pybind/mgr/telemetry/module.py [hunk 1]
```diff
@@ -403,8 +404,26 @@ def compile_report(self, channels=[]):
...
+            num_mds = 0
+            for fsm in fs_map['filesystems']:
+                fs = fsm['mdsmap']
+                report['fs']['filesystems'].append({
+                    'max_mds': fs['max_mds'],
+                    'ever_allowed_features': fs['ever_allowed_features'],
+                    'explicitly_allowed_features': fs['explicitly_allowed_features'],
+                    'in': len(fs['in']),
+                    'up': len(fs['up']),
+                    'num_mds': len(fs['info']),
```
Comment (batrick) [inline src/pybind/mgr/telemetry/module.py]: This implicitly includes the number of standby-replay daemons which I was going to ask to include anyway. Would perhaps be a good idea to explicitly calculate that in the JSON though.

File: src/pybind/mgr/telemetry/module.py [hunk 2]
```diff
@@ -403,8 +404,26 @@ def compile_report(self, channels=[]):
...
+                    'balancer_enabled': len(fs['balancer']) > 0,
+                    'num_data_pools': len(fs['data_pools']),
```
Comment (batrick) [inline src/pybind/mgr/telemetry/module.py]: `standby_count_wanted` would be useful information; i.e. how many users ever bother setting it.

Comment (batrick) [reply src/pybind/mgr/telemetry/module.py]: Also, file system age?

File: src/pybind/mgr/telemetry/module.py [hunk 3]
```diff
@@ -403,8 +404,26 @@ def compile_report(self, channels=[]):
...
+                })
+                num_mds += len(fs['info'])
```
Comment (batrick) [inline src/pybind/mgr/telemetry/module.py]: Also, if you look at per-MDS metadata, we could include number of caps and number of dentry/inodes cached.

**Comment by dillaman** [comment]
@liewegas For RBD stats:

__rbd snapshots used__: we don't keep a running total so you would technically need to open each image to count them up (since I think the PG snap map is pruned now).

__rbd num volumes__: it's probably just easiest to use the rbd python library and run a `sum(1 for _ in rbd.RBD().list2(ioctx))` for each pool flagged w/ the RBD application. If you plan to open each image to count its snapshots for above, it might also be nice to record how many images are still using the long-since deprecated v1 format (`image.old_format()`).

__rbd mirroring used__: same. you can use `rbd.RBD().mirror_mode_get(ioctx) != rbd.RBD_MIRROR_MODE_DISABLED` on each pool to collect a count. Again, if you are opening each image, you can query to see if each particular image has mirroring enabled to get a count via `image.mirror_image_get_info()['state'] != rbd.RBD_MIRROR_IMAGE_DISABLED` and

**Comment by liewegas** [comment]
@theanalyst I have everything except RGW covered now.  Is that something you could take a look at?

**Comment by liewegas** [comment]
@dillaman i opted for pool-level metadata only, and i logged the actual mirror_mode (vs a true/false), so that it will still be useful when it grows future mode values... is that okay?

**Comment by dillaman** [comment]
> @dillaman i opted for pool-level metadata only, and i logged the actual mirror_mode (vs a true/false), so that it will still be useful when it grows future mode values... is that okay?

Actually, we are planning to deprecate the mirror pool mode since (1) it was confusing to users and (2) it will conflict w/ snapshot-based mirroring. Going forward, the plan is to basically just "enable" mirroring on a pool (equivalent to today's 'image' mode), and then enable either journal or snapshot-based mirroring on a per-image basis.

**Comment by sebastian-philipp** [comment]
Is there any way to scrape Python Tracebacks from the logs?

**Comment by liewegas** [comment]
> Is there any way to scrape Python Tracebacks from the logs?

Maybe, but I think a better approach would probably be to have an exception handler that directly creates a well-formed crash record?

**Comment by liewegas** [comment]
> wrt rgw, I'm not sure whether we can determine which apis are used, sure there is a configurable rgw_apis_enabled, but unless someone explicitly disables apis, by default we enable everything, same goes for auth mechanisms (though looking at ceph.conf for rgw and specific configurables like keystone/ldap uri might be a way, not sure if rgw's config is actually accessible to the mgr other than what we register for the service?)

I think for both of these things we should do something more like adding rgw perfcounters to indicate which features are in active use.  The mgr modules have easy access to those.

**Comment by theanalyst** [comment]
added something super basic 

```diff --git a/src/pybind/mgr/telemetry/module.py b/src/pybind/mgr/telemetry/module.py
index 911a2077dd..d539f4dbca 100644
--- a/src/pybind/mgr/telemetry/module.py
+++ b/src/pybind/mgr/telemetry/module.py
@@ -561,6 +561,35 @@ class Module(MgrModule):
             report['services'] = defaultdict(int)
             for key, value in service_map['services'].items():
                 report['services'][key] += 1
+                if key == 'rgw':
+                    report['rgw'] = dict()
+                    report['rgw']['count'] = 0
+                    zones = set()
+                    realms = set()
+                    zonegroups = set()
+                    frontends = set()
+                    d = value.get('daemons', dict())
+
+                    for k,v in d.items():
+                        if k == 'summary' and v:
+                            report['rgw'][k] = v
+                        elif isinstance(v, dict) and 'metadata' in v:
+                            report['rgw']['count'] += 1
+                            zones.add(v['metadata']['zone_id'])
+                            zonegroups.add(v['metadata']['zonegroup_id'])
+                            frontends.add(v['metadata']['frontend_type#0'])
+
+                            # we could actually iterate over all the keys of
+                            # the dict and check for how many frontends there
+                            # are, but it is unlikely that one would be running
+                            # more than 2 supported ones
+                            f2 = v['metadata'].get('frontend_type#1', None)
+                            if f2:
+                                frontends.add(f2)
+
+                    report['rgw']['zones'] = len(zones)
+                    report['rgw']['zonegroups'] = len(zonegroups)
+                    report['rgw']['frontends'] = list(frontends)  # sets aren't json-serializable

```
that gives me 
```  "rgw": {
        "count": 2,
        "zones": 1,
        "zonegroups": 1,
        "frontends": [
            "beast",
            "civetweb"
        ]
    },
```

While technically this only fishes zones and zonegroups of actively running rgws, we can't actually list zones from other clusters other than having the period info (which is not yet exported to the mgr)

**Comment by liewegas** [comment]
@theanalyst I added a patch with your change, thanks!

@cbodley @mattbenjamin The other items that seem interesting include

- which auth mechanisms are in use
- which APIs (S3, swift) are in use
- how many zones
- which multisite plugins/features are used

It's really up to you what you want to learn about deployed clusters.  I think the easiest way to get this information to the mgr so that it can be included is to dump it into the service status.

**Comment by l-mb** [comment]
Woho, great. Would it be possible - this has just occurred to me - to include features used in the crushmap? Say, used bucket types and tunables?
(I'm curious how many folks use anything but straw(2), and how much data is behind those ...)

**Comment by liewegas** [comment]
Added crush info!

**Comment by liewegas** [comment]
Fine in QA.  Can I get a review so this can be merged?

**Pull request merged**